compiler/rustc_middle/src/ty/assoc.rs

```diff
@@ -25,6 +26,7 @@ pub struct AssocItem {
 
     /// If this is an item in an impl of a trait then this is the `DefId` of
```

Comment:
The comment is incorrect, this field is currently set to `Some` for `AssocItemContainer::Trait` as well, but not always (?!)

It should either be consistently set for all `AssocItemContainer::Trait`s, or not set for any of them.
In the latter case the conditions `container == AssocItemContainer::Trait && trait_item_def_id.is_some()` will become redundant.

Reply:
> The comment is incorrect, this field is currently set to `Some` for `AssocItemContainer::Trait` as well, but not always (?!)

It seems to me somebody decided to put `Some` there cause it's "free", but it's not the purpose of the field. I could fix this, maybe in a separate PR? Or I could fix the code comment.

> `container` is almost a duplicate of `trait_item_def_id`, except for error recovery cases

I noticed this as well. This change makes the two things closer to being redundant, but they are still not totally redundant. So ultimately I think this change is good and I don't see a way to simplify it.

Reply:
I'm talking about this because changing representation of `AssocItem` would require rewriting half of the changes in this PR, so it makes sense to not separate them.
(The HIR change can be landed separately though.)

The non-redundant representation for `AssocItem` would be
```rust
struct AssocItem {
    def_id: DefId,
    kind: AssocKind,
    other_kind: OtherKind,
}

enum OtherKind {
    Trait,
    InherentImpl,
    TraitImpl(Option<DefId>),
}
```

Reply:
Gotcha. I can give that I try.